src/lsp4clj/server.clj

```diff
@@ -330,7 +349,7 @@
         (if-let [{:keys [p started] :as req} (get pending-requests id)]
           (do
             (trace this trace/received-response req resp started now)
-            (deliver p (if error resp result)))
+            (p/resolve! p (if error resp result)))
```

Comment:
If I understand you correctly we do not `reject!` on error because the former implementation did not support rejected promises, and we don't want to break existing code.

Should we document this here?

Maybe we could add an option to `send-request` to throw exceptions on error, similar to how `clj-http` supports `:throw-exceptions? false`. Then new code could set this to `true` and eventually we could make this the default?

Reply:
> we do not reject! on error because the former implementation did not support rejected promises, and we don't want to break existing code

Correct, existing code expects to `(let [{:keys [error] :as resp} (deref-or-cancel req ...`.

> Maybe we could add an option to `send-request` to throw exceptions on error

You mean reject on error, so that dereffing would throw? That might be nice, although I don't want to add it in this PR. Something like that could be a per-request setting, as part of `send-request`, or it could be a global setting, as part of `chan-server`.

> Should we document this here?

Yes, I'll at least add some documentation.

Reply:
**Edit:** comments added.

Reply:
I'm going to leave this as is for now. It'd be good to revisit in a v2, or I'd be happy to review a PR.

> I, personally, would treat PendingRequest as a promesa promise after this PR is merged, and stop using the old interface.

As soon as you convert the PendingRequest to a promesa promise, you lose access to `deref-or-cancel`. To replicate that, you'd have a few options.
* You could write your own code that added a timeout to the promise, caught the timeout, and did a cancellation.
* You could add more code to lsp4clj that encapsulated that pattern, although you'd basically be back to where we are now.

Reply:
I think implementing `deref-or-cancel` for `CompletableFuture` is trivial, and could be provided by `lsp4clj`. This could be a step towards getting rid of `PendingRequest` altogether.

To be honest, but this might be a different topic, I'm not convinced that `deref-or-cancel` should be used anyway. When we do not get a response in time, chances are high that there was a problem with the original request, and I'm not sure if the client would actually benefit from knowing that the server is no longer interested in the response. Do we have evidence from clojure-lsp that this is helpful?

Sending a cancel actually adds congestion, as we will block on the output channel to send the cancel notification. We could just as well go on and ignore any response that comes in later.

Reply:
> I think implementing deref-or-cancel for CompletableFuture is trivial... This could be a step towards getting rid of PendingRequest altogether.

Yeah, sounds like a nice addition. I've run out of steam in this PR though. :)

> We could just as well go on and ignore any response that comes in later.

We do ignore the response, whether we timeout using a `deref` with a timeout, or with `deref-or-cancel`.

> I'm not sure if the client would actually benefit from knowing that the server is no longer interested in the response. Do we have evidence from clojure-lsp that this is helpful?

The server doesn't benefit from sending the cancellation. It's a courtesy to the client, potentially allowing it to skip or abort some work. I agree, it's unclear whether the client really benefits from that though. It's possible things are so broken on the client that a single cancellation won't help it. On the other hand, lsp4clj is a low-level tool, and so it seems like it should stay close to the spec. The spec suggests that the server should _try_ to send a cancellation, so I think it's a facility lsp4clj should offer. Of course higher level tools like clojure-lsp and Scarlet are free to use `deref-or-cancel`, or `deref` with a timeout, as they see fit.

clojure-lsp certainly sees benefit from the client sending it cancellations, but that's largely because the client tends to hammer the server with repeated requests. For example, VS Code is very eager about asking for code actions (gotta keep that 💡  up to date!). It seems to ask for them with nearly every key press. But, as a user keeps typing, it cancels earlier requests, which really does let the server skip some work.

clojure-lsp doesn't initiate many requests, and AFAIK never sends many requests right in a row. I think the only time we do send several in a row is to create a document and then open it. Scarlet does occasionally send several requests in a row, and the client does occasionally have trouble processing all of those and so Scarlet sometimes times out waiting for a response. The server sends a cancellation because it uses `deref-or-cancel`, but I'm not aware of any cases that actually hurts either the client or server.

Anyway, I can see it both ways. 

> Sending a cancel actually adds congestion, as we will block on the output channel to send the cancel notification. 

That's technically true, but I don't see it as being a big problem in practice. The only way we end up blocking on the output channel is if the output stream blocks. But there are several layers of buffers that prevent that from happening—our process, the OS, and the client all have buffers to avoid blocking that stream. If the client has stopped or slowed down reading, and all those buffers fill up, then I think we have bigger problems to fix.

I'll certainly admit that you have better instincts about the potential problems here. This project has lots of concurrency pitfalls. I'm not sure whether that's an architectural problem or it's inherent in the problem space. Anyway, it's worth more effort at some point.

Reply:
> Yeah, sounds like a nice addition. I've run out of steam in this PR though. :)

Yeah, sorry for opening up a completely different topic. I'm happy to help, if you want?

Let's put a pin in the `deref-or-cancel` discussion for now.

My only remaining concern is that we now have the chance to return "idiomatic" promises, and with that I mean reject on error, instead of resolving with `result`. I think it is confusing to users who will use the promesa interface of `PendingRequest` but need to implement custom error handling. If we create facts now, we might be hesitant in the future to change this.

I therefore suggest to `p/reject!` the promise on error here:

```suggestion
            (if error
              (p/reject! p (ex-info "Received error response" resp))
              (p/resolve! p result))
```

For backward compatibility, `PendingRequest`'s `deref` and `deref-or-cancel` could use a helper function like this:

```clojure
(defn- resolve-ex-info [p]
  (p/catch (fn [ex] (or (ex-data ex) (p/rejected ex)))))
```

Calls to `(deref p ,,,)` then change to: `(deref (resolve-ex-info p) ,,,)`.

This way, existing could can work as is, but async code can use the promise API and get what they expect. We should document that `(deref pending-request)` and `(deref (p/promise pending-request))` report errors differently, but I think this is worth it.

Reply:
@ferdinand-beyer sorry for dropping the ball on this. For some reason I thought that we had merged some part of this PR already. Unfortunately, I just merged a Scarlet PR that assumes that this lsp4clj PR has been merged. So now I'm anxious to get this PR merged and released.

I've made the change you suggested. It now seems possible to use the request object as a promesa promise fairly idiomatically. If you'd review the latest commit to ensure it's what you envisioned, I'd appreciate it.

Before you done that, I'm going to start working with @ericdallo to get this merged. @ericdallo would you also do a review. If you'd like I can remind myself how to merge and do a release, or if you'd be willing to do that, I'd appreciate it.

Reply:
Sure, I'll take a look again!